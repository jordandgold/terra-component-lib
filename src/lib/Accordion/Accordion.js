import React, { Component } from "react";
import Icon from "../Icon/Icon";
import "./Accordion.scss";
import PropTypes from "prop-types";

class Accordion extends Component {
  constructor(props) {
    super(props);

    this.state = {
      active: undefined
    };
  }

  checkActive = () => {
    if (this.props.defaultActive === 0 || this.props.defaultActive) {
      this.setState({
        active: this.props.defaultActive
      });
    }
  };

  accordionToggle = fold => {
    if (this.state.active === fold) {
      this.setState({
        active: undefined
      });
    } else {
      this.setState({ active: fold });
    }
  };

  generateFolds = () => {
    return this.props.children.map((child, index) => {
      let activeClass = index == this.state.active ? "is-expanded" : "";
      return (
        <div
          key={`accordion-fold-${index}`}
          className={`ter-accordion__fold ${activeClass}`}
        >
          <button
            className="ter-accordion__trigger"
            onClick={() => this.accordionToggle(index)}
          >
            {child.props.title}
            <Icon
              type="open-caret-right-dark-16px"
              className="ter-accordion__trigger-caret"
              size="16px"
            />
          </button>
          {child}
        </div>
      );
    });
  };

  componentDidMount() {
    this.checkActive();
  }

  render() {
    return <div className="ter-accordion">{this.generateFolds()}</div>;
  }
}

export default Accordion;

export class AccordionFold extends Component {
  render() {
    return (
      <React.Fragment>
        <div className="ter-accordion__content">{this.props.children}</div>
      </React.Fragment>
    );
  }
}

Accordion.propTypes = {
  defaultActive: PropTypes.number,
  children: PropTypes.node.isRequired
};
