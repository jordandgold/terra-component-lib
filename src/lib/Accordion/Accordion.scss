@import "../global-styles/terra.scss";

.ter-accordion {
  margin-bottom: spacing("sm");
  border-bottom: $accordion-border-width solid $accordion-border-color;
  &__fold {
    font-size: $accordion-font-size;
    margin-bottom: spacing("sm");
    border-color: $accordion-border-color;
    border-width: $accordion-border-width;
    border-style: solid;
    border-left-width: 0px;
    border-right-width: 0px;
    transition: background 225ms ease-in-out;
    margin-bottom: 0;
    border-bottom: 0px;
    &:first-child {
      border-bottom: 0px;
    }
    &:last-child {
      border-bottom: 0px;
    }
  }
  &__trigger {
    @include button-reset();
    font-size: $accordion-font-size;
    padding: spacing("xs") spacing("lg") spacing("xs") spacing("sm");
    position: relative;
    display: block;
    text-decoration: none;
    font-weight: font-weight("bold");
    width: 100%;
    text-align: left;
    line-height: line-height("loose");
    &-caret {
      position: absolute;
      color: $accordion-arrow-color;
      top: 50%;
      right: spacing("sm");
      transform: rotate(0deg) translateY(-50%);
      transition: transform 200ms ease-in-out;
      font-size: 16px;
      width: 16px;
      height: 16px;
    }
    &:hover {
      cursor: pointer;
      text-decoration: none;
    }
  }
  &__content {
    font-size: $accordion-font-size;
    display: none;
    padding: spacing("xs") spacing("sm");
    p {
      font-size: $accordion-font-size;
    }
  }
  &__fold {
    &.is-expanded {
      .ter-accordion__content {
        display: block;
      }
      .ter-accordion__trigger {
        text-decoration: none;
        &-caret {
          transform: rotate(90deg) translateX(-50%);
        }
      }
    }
  }
}
