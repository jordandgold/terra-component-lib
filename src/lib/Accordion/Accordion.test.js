import React from "react";
import { shallow } from "enzyme";
import Accordion, { AccordionFold } from "./Accordion";

describe("Accordion", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Accordion defaultActive={0}>
        <p>test fold</p>
        <p>test fold 2</p>
      </Accordion>
    );
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("checkActive", () => {
    it("should setState if conditions are met", () => {
      wrapper.setState({ active: 10 });
      wrapper.instance().checkActive();

      expect(wrapper.state().active).toEqual(0);
    });

    it("should setState if conditions are met", () => {
      wrapper = shallow(
        <Accordion defaultActive={2}>
          <p>test fold</p>
          <p>test fold 2</p>
        </Accordion>
      );
      wrapper.state().active = 10;
      wrapper.instance().checkActive();

      expect(wrapper.state().active).toEqual(2);
    });

    it("should not call setState if conditions arent met", () => {
      wrapper = shallow(
        <Accordion>
          <p>test fold</p>
          <p>test fold 2</p>
        </Accordion>
      );
      wrapper.instance().checkActive();

      expect(wrapper.state().active).toEqual(undefined);
    });
  });

  describe("accordionToggle", () => {
    it("should set state to undefined if active = fold", () => {
      wrapper.state().active = "hello";

      wrapper.instance().accordionToggle("hello");

      expect(wrapper.state().active).toEqual(undefined);
    });

    it("should otherwise set active to argument", () => {
      wrapper.instance().accordionToggle("hello");

      expect(wrapper.state().active).toEqual("hello");
    });
  });

  describe("generateFold", () => {
    it("should return some JSX", () => {
      const expected = [
        '<div class="ter-accordion__fold is-expanded"><button class="ter-accordion__trigger"><svg viewBox="0 0 8 8" class="ter-icon ter-icon--open-caret-right-dark-16px ter-icon--16px ter-accordion__trigger-caret"><use xlink:href="[object Object]#ter-icon--open-caret-right-dark-16px"></use></svg></button><p>test fold</p></div>',
        '<div class="ter-accordion__fold "><button class="ter-accordion__trigger"><svg viewBox="0 0 8 8" class="ter-icon ter-icon--open-caret-right-dark-16px ter-icon--16px ter-accordion__trigger-caret"><use xlink:href="[object Object]#ter-icon--open-caret-right-dark-16px"></use></svg></button><p>test fold 2</p></div>'
      ];

      const result = wrapper
        .instance()
        .generateFolds()
        .map(fold => {
          return shallow(fold).html();
        });

      expect(result).toEqual(expected);
    });

    it("should call generateFold onClick", () => {
      const spy = jest.spyOn(wrapper.instance(), "generateFolds");
      wrapper.instance().forceUpdate();

      wrapper
        .find(".ter-accordion__fold:first-child .ter-accordion__trigger")
        .simulate("click");

      expect(spy).toHaveBeenCalled();
    });
  });

  describe("AccordionFold", () => {
    let foldWrapper;

    beforeEach(() => {
      foldWrapper = shallow(<AccordionFold children={"<p>hello</p>"} />);
    });

    it("should match the snapshot", () => {
      expect(foldWrapper).toMatchSnapshot();
    });
  });
});
