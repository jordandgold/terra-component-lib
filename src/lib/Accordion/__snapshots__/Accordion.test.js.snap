// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion AccordionFold should match the snapshot 1`] = `
<Fragment>
  <div
    className="ter-accordion__content"
  >
    &lt;p&gt;hello&lt;/p&gt;
  </div>
</Fragment>
`;

exports[`Accordion should match the snapshot 1`] = `
<div
  className="ter-accordion"
>
  <div
    className="ter-accordion__fold is-expanded"
    key="accordion-fold-0"
  >
    <button
      className="ter-accordion__trigger"
      onClick={[Function]}
    >
      <Icon
        className="ter-accordion__trigger-caret"
        size="16px"
        type="open-caret-right-dark-16px"
      />
    </button>
    <p>
      test fold
    </p>
  </div>
  <div
    className="ter-accordion__fold "
    key="accordion-fold-1"
  >
    <button
      className="ter-accordion__trigger"
      onClick={[Function]}
    >
      <Icon
        className="ter-accordion__trigger-caret"
        size="16px"
        type="open-caret-right-dark-16px"
      />
    </button>
    <p>
      test fold 2
    </p>
  </div>
</div>
`;
